src/modules/uORB/uORBDeviceNode.cpp

```diff
@@ -130,18 +172,19 @@ uORB::DeviceNode::copy(void *dst, unsigned &generation)
 			ATOMIC_ENTER;
 			const unsigned current_generation = _generation.load();
 
-			if (current_generation > generation + _queue_size) {
-				// Reader is too far behind: some messages are lost
-				generation = current_generation - _queue_size;
-			}
-
-			if ((current_generation == generation) && (generation > 0)) {
+			if (current_generation == generation && current_generation > 0) {
```

Comment:
As a small optimization this could be a special case that directly does the simple memcpy (ignoring is_in_range()) and returns.

Reply:
Yes, it is possible, but I think that program structure is not very good.

Let me explain my question:
> The current construction _data and _generation.fetch_add(1) are not completed by the same atomic operation. Can we change it?

I originally wanted to delete the judgment that `current_generation > 0`, so there is no special condition judgment.

We only construct `_data` when we publish for the first time, and then we publish the `_data`, `_generation++`. But the construction of `_data` and `_generation++` are not completed under the same atomic operation.
https://github.com/PX4/Firmware/blob/master/src/modules/uORB/uORBDeviceNode.cpp#L193

If they are completed under the same atomic operation, then It can be guaranteed that when `_data` is valid, `_generation` must be valid, and there is no need to judge that it is greater than 0.

Reply:
That's a tempting clean solution, but actually I think we need to go a bit in the opposite direction and allocate `_data` alongside the uORB::DeviceNode. From a safety perspective the general rule is to not allocate anything in flight, and there are still a number of things that won't publish until operating (eg entering mission).

Reply:
Now I think we don't care about the condition of `current_generation> 0`. If there is no published, all the data in `_data` is invalid, so it doesn't matter which index is copied.

After `generation--` it becomes `UINT_MAX`, but it doesn't matter, all data is invalid now, and then `generation++` is executed.

I will delete it.